modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/CursorSubscription.java

```diff
@@ -0,0 +1,185 @@
+    private boolean isDone = false;
+
+    /** Cached response from the Meta Storage leader. {@code null} if no requests have been issued yet. */
+    @Nullable
+    private BatchResponse cachedResponse;
+
+    /** Index inside the cached response. */
+    private int responseIndex;
+
+    /** Amount of entries requested by the subscriber. */
+    private long demand;
+
+    CursorSubscription(
+            RaftGroupService raftService,
+            MetaStorageCommandsFactory commandsFactory,
+            ExecutorService executorService,
+            IgniteUuid cursorId,
+            Subscriber<? super Entry> subscriber
+    ) {
+        this.raftService = raftService;
+        this.commandsFactory = commandsFactory;
+        this.executorService = executorService;
+        this.cursorId = cursorId;
+        this.subscriber = subscriber;
+    }
+
+    @Override
+    public void request(long n) {
+        if (n <= 0) {
+            onError(new IllegalArgumentException("Requested amount must be greater than zero, got: " + n));
+
+            return;
+        }
+
+        if (isDone) {
+            return;
+        }
+
+        // Increase the number of requested items.
+        demand += n;
+
+        if (demand <= 0) {
+            onError(new IllegalArgumentException("Long overflow"));
+
+            return;
+        }
+
+        // Start the processing if it has not been started yet (async operation).
+        if (cachedResponse == null) {
+            requestNextBatch();
+        }
+    }
+
+    /**
+     * Does the actual request processing. Must always be called on {@link CursorSubscription#executorService}.
+     */
+    private void processRequest() {
+        if (isDone) {
+            return;
+        }
+
+        try {
+            assert cachedResponse != null;
+
+            List<Entry> entries = cachedResponse.entries();
+
+            while (demand > 0 && !isDone) {
+                if (responseIndex < entries.size()) {
+                    subscriber.onNext(entries.get(responseIndex));
+
+                    responseIndex++;
+                    demand--;
+                } else {
+                    if (cachedResponse.hasNextBatch()) {
+                        requestNextBatch();
+                    } else {
+                        isDone = true;
+
+                        subscriber.onComplete();
+                    }
+
+                    return;
+                }
+            }
+        } catch (Exception e) {
+            onError(e);
+        }
+    }
+
+    private void requestNextBatch() {
+        NextBatchCommand command = commandsFactory.nextBatchCommand()
+                .cursorId(cursorId)
+                .batchSize(BATCH_SIZE)
+                .build();
+
+        raftService.<BatchResponse>run(command)
+                .whenCompleteAsync((resp, e) -> {
+                    if (e == null) {
+                        cachedResponse = resp;
```

Comment:
Not sure that I get it. Do you expect that whenCompleteAsync will be executed in the same thread as CursorSubscription.request(n)?

Reply:
> Is it always singleTreaded executor?

No

> Do you expect that whenCompleteAsync will be executed in the same thread as CursorSubscription.request(n)

No, but I expect that there's a following HB relationships:
1. For the first ever call of `request`: request -> submission on the executor. Though now I think that this may be debatable, because we are completing the future from a Raft service thread. Looking at the code I still think that there's a HB there because of the future chains, but maybe it makes sense to make `demand` variable volatile, what do you think?
2. For all consecutive calls `request` is expected to be done in the same thread as the `processRequest`  (because it is probably called inside the `onNext` callback)

Reply:
My fault, I've missed the 
```
        if (cachedResponse == null) {
            requestNextBatch();
        }
```
part. So you'll never request next portion before current one is fully processed, so there will be no multiple requestNextBatch calls on different threads.

Seems that demand should be fine without being volatile.

Reply:
Correct, there should be no concurrent `requestNextBatch` at the same moment, though consecutive batches can be processed on different threads (and I expect that we have a HB through future chains there)